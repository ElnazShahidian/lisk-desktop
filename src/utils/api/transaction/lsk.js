/* eslint-disable max-lines */
import { transactions } from '@liskhq/lisk-client';

import {
  tokenMap,
  minFeePerByte,
  DEFAULT_NUMBER_OF_SIGNATURES,
  DEFAULT_SIGNATURE_BYTE_SIZE,
  MODULE_ASSETS_MAP,
  moduleAssetSchemas,
  BASE_FEES,
} from '@constants';

import { joinModuleAndAssetIds } from '@utils/moduleAssets';
import { createTransactionObject } from '@utils/transaction';
import { validateAddress } from '../../validators';
import http from '../http';
import { getDelegates } from '../delegate';

const httpPrefix = '/api/v2';

const httpPaths = {
  fees: `${httpPrefix}/fees`,
  transactions: `${httpPrefix}/transactions`,
  transaction: `${httpPrefix}/transactions`,
  transactionStats: `${httpPrefix}/transactions/statistics`,
  schemas: `${httpPrefix}/transactions/schemas`,
};

/**
 * Retrieves the details of a single transaction
 *
 * @param {Object} data
 * @param {String} data.params
 * @param {String} data.params.id - Id of the transaction
 * @param {String?} data.baseUrl - Lisk Service API url to override the
 * existing ServiceUrl on the network param. We may use this to retrieve
 * the details of an archived transaction.
 * @param {Object} data.network - Network setting from Redux store
 * @returns {Promise} Transaction details API call
 */
export const getTransaction = ({
  params, network, baseUrl,
}) => http({
  path: httpPaths.transaction,
  params,
  network,
  baseUrl,
});

const filters = {
  address: { key: 'address', test: address => !validateAddress(tokenMap.LSK.key, address) },
  timestamp: { key: 'timestamp', test: str => /^(\d+)?:(\d+)?$/.test(str) },
  amount: { key: 'amount', test: str => /^(\d+)?:(\d+)?$/.test(str) },
  limit: { key: 'limit', test: num => parseInt(num, 10) > 0 },
  offset: { key: 'offset', test: num => parseInt(num, 10) >= 0 },
  moduleAssetId: { key: 'moduleAssetId', test: str => /\d:\d/.test(str) },
  height: { key: 'height', test: num => parseInt(num, 10) > 0 },
  sort: {
    key: 'sort',
    test: str => ['amount:asc', 'amount:desc', 'fee:asc', 'fee:desc', 'timestamp:asc', 'timestamp:desc'].includes(str),
  },
};

/**
 * Retrieves the list of transactions for given parameters
 *
 * @param {Object} data
 * @param {Object} data.network - Network setting from Redux store
 * @param {String?} data.baseUrl - Lisk Service API url to override the
 * existing ServiceUrl on the network param. We may use this to retrieve
 * the details of an archived transaction.
 * @param {Object} data.params
 * @param {String} data.params.address Sender or recipient account
 * @param {String} data.params.dateFrom Unix timestamp, the start time of txs
 * @param {String} data.params.dateTo Unix timestamp, the end time of txs
 * @param {String} data.params.amountFrom The minimum value of txs
 * @param {String} data.params.amountTo The maximum value of txs
 * @param {String} data.params.moduleAssetId The moduleAssetId. 2:0, 5:1, etc
 * @param {Number} data.params.offset Used for pagination
 * @param {Number} data.params.limit Used for pagination
 * @param {String} data.params.sort an option of 'amount:asc',
 * 'amount:desc', 'timestamp:asc', 'timestamp:desc',
 * @param {Object} data.params.height The height of the block whose transaction we want.
 * If passed, all other parameter will be ignored.
 * @returns {Promise} Transactions list API call
 */
export const getTransactions = ({
  network,
  params,
  baseUrl,
}) => {
  const normParams = {};

  // if blockId, ignore others
  if (params.blockId) {
    normParams.block = params.blockId;
  } else {
    // Validate params and fix keys
    Object.keys(params).forEach((key) => {
      if (filters[key] && filters[key].test(params[key])) {
        normParams[filters[key].key] = params[key];
      } else {
        // eslint-disable-next-line no-console
        console.log(`getTransactions: Dropped ${key} parameter, it's invalid.`, params[key]);
      }
    });
  }

  return http({
    network,
    path: httpPaths.transactions,
    params: normParams,
    baseUrl,
  });
};

/**
 * Fetches and generates an array of monthly number of delegate
 * registrations on Lisk blockchain.
 *
 * @param {Object} Network - Network setting from Redux store
 * @returns {Promise} Registered delegates list API call
 */
export const getRegisteredDelegates = async ({ network }) => {
  const delegates = await getDelegates({
    network,
    params: { limit: 1 },
  });
  const txs = await getTransactions({
    network,
    params: { moduleAssetId: '5:0', limit: 100 },
  });

  if (delegates.error || txs.error) {
    return Error('Error fetching data.');
  }

  const getDate = (timestamp) => {
    const d = new Date(timestamp * 1000);
    return `${new Date(d).getFullYear()}-${new Date(d).getMonth() + 1}`;
  };

  // create monthly number of registration as a dictionary
  const monthStats = txs.data
    .map(tx => tx.block.timestamp)
    .reduce((acc, timestamp) => {
      const date = getDate(timestamp);
      acc[date] = typeof acc[date] === 'number' ? acc[date] + 1 : 1;
      return acc;
    }, {});

  // Create a sorted array of monthly accumulated number of registrations
  const res = Object.keys(monthStats)
    .sort((a, b) => -1 * (b - 1))
    .reduce((acc, month) => {
      if (acc[0][0] === month) {
        acc.unshift([null, acc[0][1] - monthStats[month]]);
      } else if (acc[0][0] === null) {
        acc[0][0] = month;
        acc.unshift([null, acc[0][1] - monthStats[month]]);
      }

      return acc;
    }, [[getDate(txs.data[0].block.timestamp), delegates.meta.total]]);

  // Add the date of one month before the last tx
  res[0][0] = getDate(txs.data[txs.data.length - 1].block.timestamp - 2670000);

  return res;
};

/**
 * Retrieves the overall statistics of network transactions.
 *
 * @param {Object} data
 * @param {Object} data.params
 * @param {String} data.params.period - An option of 'day' or 'month'
 * @param {Number} data.params.limit - The number of results
 * @param {Object} data.network - Network setting from Redux store
 * @returns {Object} Network transactions statistics
 */
export const getTransactionStats = ({ network, params: { period } }) => {
  const normParams = {
    week: { path: 'day', limit: 7 },
    month: { path: 'month', limit: 6 },
    year: { path: 'month', limit: 12 },
  };

  return http({
    path: `${httpPaths.transactionStats}/${normParams[period].path}`,
    params: { limit: normParams[period].limit },
    network,
  });
};

/**
 * Retrieves transaction schemas.
 *
 * @param {Object} data
 * @param {String?} data.baseUrl - Lisk Service API url to override the
 * existing ServiceUrl on the network param. We may use this to retrieve
 * the details of an archived transaction.
 * @param {Object} data.network - Network setting from Redux store
 * @returns {Promise} http call
 */
export const getSchemas = ({ baseUrl }) => http({
  path: httpPaths.schemas,
  baseUrl,
});

/**
 * Returns a dictionary of base fees for low, medium and high processing speeds
 * @returns {Promise<{Low: number, Medium: number, High: number}>} with low,
 * medium and high priority fee options
 */
export const getTransactionBaseFees = network =>
  http({
    path: httpPaths.fees,
    searchParams: {},
    network,
  })
    .then((response) => {
      const { feeEstimatePerByte } = response.data;
      return {
        Low: feeEstimatePerByte.low,
        Medium: feeEstimatePerByte.medium,
        High: feeEstimatePerByte.high,
      };
    });

/**
 * Returns the actual tx fee based on given tx details
 * and selected processing speed
 *
 * @param {String} txData - The transaction object
 * @param {Object} selectedPriority - network configuration
 * @returns {Promise} Object containing value, error and feedback
 */
// eslint-disable-next-line max-statements
export const getTransactionFee = async ({
  transaction, selectedPriority,
}) => {
  const feePerByte = selectedPriority.value;

  const {
    moduleAssetId, ...rawTransaction
  } = transaction;
  const schema = moduleAssetSchemas[moduleAssetId];
  const maxAssetFee = MODULE_ASSETS_MAP[moduleAssetId].maxFee;
  const transactionObject = createTransactionObject(rawTransaction, moduleAssetId);

  const minFee = transactions.computeMinFee(schema, {
    ...transactionObject,
    signatures: undefined,
  }, {
    baseFees: BASE_FEES,
  });

  // tie breaker is only meant for medium and high processing speeds
  const tieBreaker = selectedPriority.selectedIndex === 0
    ? 0 : minFeePerByte * feePerByte * Math.random();

  const size = transactions.getBytes(schema, {
    ...transactionObject,
    signatures: new Array(DEFAULT_NUMBER_OF_SIGNATURES).fill(
      Buffer.alloc(DEFAULT_SIGNATURE_BYTE_SIZE),
    ),
  }).length;

  const calculatedFee = Number(minFee + BigInt(size * feePerByte) + BigInt(tieBreaker));
  const fee = Math.min(calculatedFee, maxAssetFee);
  const roundedValue = transactions.convertBeddowsToLSK(fee.toString());

  const feedback = transaction.amount === ''
    ? '-'
    : `${(roundedValue ? '' : 'Invalid amount')}`;

  return {
    value: roundedValue,
    error: !!feedback,
    feedback,
  };
};

/**
 * creates a new transaction
 *
 * @param {Object} transaction The transaction information
 * @param {String} transactionType The transaction type title
 * @returns {Promise} promise that resolves to a transaction or
 * rejects with an error
 */
export const create = ({
  network,
  moduleAssetId,
  ...transactionObject
}) => new Promise((resolve, reject) => {
  const { networkIdentifier } = network.networks.LSK;
  const {
    passphrase, ...rawTransaction
  } = transactionObject;

  const schema = moduleAssetSchemas[moduleAssetId];
  const transaction = createTransactionObject(rawTransaction, moduleAssetId);

  try {
    const signedTransaction = transactions.signTransaction(
      schema, transaction, Buffer.from(networkIdentifier, 'hex'), passphrase,
    );

    resolve(signedTransaction);
  } catch (error) {
    reject(error);
  }
});

/**
 * broadcasts a transaction over the network
 *
 * @param {object} transaction
 * @param {Object} network
 * @param {string} network.name - the network name, e.g. mainnet, betanet
 * @param {string} network.address - the node address e.g. https://betanet-lisk.io
 * @returns {Promise} promise that resolves to a transaction or rejects with an error
 */
export const broadcast = async ({ transaction, serviceUrl }) => {
  const moduleAssetId = joinModuleAndAssetIds({
    moduleID: transaction.moduleID,
    assetID: transaction.assetID,
  });
  const schema = moduleAssetSchemas[moduleAssetId];
  const binary = transactions.getBytes(schema, transaction);
  const payload = binary.toString('hex');
  const body = JSON.stringify({ transaction: payload });

  const response = await http({
    method: 'POST',
    baseUrl: serviceUrl,
    path: '/api/v2/transactions',
    body,
  });

  return response;
};
