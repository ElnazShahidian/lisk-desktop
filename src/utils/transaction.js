import {
  MODULE_ASSETS_NAME_ID_MAP,
} from '@constants';
import {
  extractAddressFromPublicKey,
  getBase32AddressFromAddress,
  getAddressFromBase32Address,
} from '@utils/account';
import { transformStringDateToUnixTimestamp } from '@utils/datetime';
import { toRawLsk } from '@utils/lsk';
import { splitModuleAndAssetIds, joinModuleAndAssetIds } from '@utils/moduleAssets';

const {
  transfer, voteDelegate, registerDelegate, unlockToken, reclaimLSK, registerMultisignatureGroup,
} = MODULE_ASSETS_NAME_ID_MAP;

const EMPTY_BUFFER = Buffer.from('');
const convertStringToBinary = value => Buffer.from(value, 'hex');
const convertBinaryToString = value => value.toString('hex');

/**
 * Converts a transaction returned by lisk elements back to the signature
 * used by lisk desktop
 * @param {object} transaction - the transaction object
 * @returns the transformed transaction
 */
// eslint-disable-next-line max-statements
const transformTransaction = ({
  moduleID, assetID, id, asset, nonce, fee, senderPublicKey, signatures,
}) => {
  const moduleAssetId = joinModuleAndAssetIds({ moduleID, assetID });
  const senderAddress = extractAddressFromPublicKey(senderPublicKey);
  const transformedTransaction = {
    moduleAssetId,
    id: convertBinaryToString(id),
    fee: String(fee),
    nonce: String(nonce),
    signatures,
    sender: {
      address: senderAddress,
      publicKey: convertBinaryToString(senderPublicKey),
    },
  };

  switch (moduleAssetId) {
    case transfer: {
      transformedTransaction.asset = {
        data: asset.data,
        amount: String(asset.amount),
        recipient: { address: getBase32AddressFromAddress(asset.recipientAddress) },
      };

      break;
    }

    case registerDelegate: {
      transformedTransaction.asset = {
        username: asset.username,
      };
      break;
    }

    case voteDelegate: {
      transformedTransaction.asset = {
        votes: asset.votes.map(vote => ({
          amount: Number(vote.amount),
          delegateAddress: getBase32AddressFromAddress(vote.delegateAddress),
        })),
      };
      break;
    }

    case reclaimLSK: {
      transformedTransaction.asset = {
        amount: asset.amount,
      };
      break;
    }

    case unlockToken: {
      transformedTransaction.asset = {
        unlockObjects: asset.unlockObjects.map(unlockingObject => ({
          delegateAddress: getBase32AddressFromAddress(unlockingObject.delegateAddress),
          amount: Number(unlockingObject.amount),
          unvoteHeight: unlockingObject.height.start,
        })),
      };
      break;
    }

    case registerMultisignatureGroup: {
      transformedTransaction.asset = {
        numberOfSignatures: asset.numberOfSignatures,
        mandatoryKeys: asset.mandatoryKeys.map(convertBinaryToString),
        optionalKeys: asset.optionalKeys.map(convertBinaryToString),
      };
      break;
    }

    default:
      throw Error('Unknown transaction');
  }

  return transformedTransaction;
};

/**
 * creates a transaction object to be used with the api client from
 * lisk elements
 * @param {object} tx - the transaction data
 * @param {string} moduleAssetId - moduleAssetId
 * @returns the transaction object
 */
// eslint-disable-next-line max-statements
const createTransactionObject = (tx, moduleAssetId) => {
  const [moduleID, assetID] = splitModuleAndAssetIds(moduleAssetId);
  const {
    senderPublicKey, nonce, amount, recipientAddress, data, fee = 0,
  } = tx;

  const transaction = {
    moduleID,
    assetID,
    senderPublicKey: convertStringToBinary(senderPublicKey),
    nonce: BigInt(nonce),
    fee: BigInt(fee),
    signatures: [],
  };

  switch (moduleAssetId) {
    case transfer: {
      const binaryAddress = recipientAddress
        ? getAddressFromBase32Address(recipientAddress) : EMPTY_BUFFER;

      transaction.asset = {
        recipientAddress: binaryAddress,
        amount: BigInt(amount),
        data,
      };

      break;
    }

    case registerDelegate: {
      transaction.asset = {
        username: tx.username,
      };
      break;
    }

    case voteDelegate: {
      const votes = tx.votes.map(vote => ({
        amount: BigInt(vote.amount),
        delegateAddress: getAddressFromBase32Address(vote.delegateAddress),
      }));
      transaction.asset = { votes };
      break;
    }

    case unlockToken: {
      transaction.asset = {
        unlockObjects: tx.unlockingObjects.map(unlockingObject => ({
          amount: BigInt(unlockingObject.amount),
          delegateAddress: getAddressFromBase32Address(unlockingObject.delegateAddress),
          unvoteHeight: unlockingObject.unvoteHeight,
        })),
      };
      break;
    }

    case reclaimLSK: {
      transaction.asset = {
        amount: BigInt(amount),
      };
      break;
    }

    case registerMultisignatureGroup: {
      transaction.asset = {
        numberOfSignatures: Number(tx.numberOfSignatures),
        mandatoryKeys: tx.mandatoryKeys.map(convertStringToBinary),
        optionalKeys: tx.optionalKeys.map(convertStringToBinary),
      };
      break;
    }

    default:
      throw Error('Unknown transaction');
  }

  return transaction;
};

const containsTransactionType = (transactions = [], type) =>
  transactions.some(tx => tx.moduleAssetId === type);

/**
 * Adapts transaction filter params to match transactions API method
 *
 * @param {Object} params - Params received from withFilters HOC
 * @returns {Object} - Parameters consumable by transaction API method
 */
const normalizeTransactionParams = params => Object.keys(params)
  .reduce((acc, item) => {
    switch (item) {
      case 'dateFrom':
        if (!acc.timestamp) acc.timestamp = ':';
        acc.timestamp = acc.timestamp.replace(/(\d+)?:/, `${transformStringDateToUnixTimestamp(params[item])}:`);
        break;
      case 'dateTo':
        if (!acc.timestamp) acc.timestamp = ':';
        acc.timestamp = acc.timestamp.replace(/:(\d+)?/, `:${transformStringDateToUnixTimestamp(params[item])}`);
        break;
      case 'amountFrom':
        if (!acc.amount) acc.amount = ':';
        acc.amount = acc.amount.replace(/(\d+)?:/, `${toRawLsk(params[item])}:`);
        break;
      case 'amountTo':
        if (!acc.amount) acc.amount = ':';
        acc.amount = acc.amount.replace(/:(\d+)?/, `:${toRawLsk(params[item])}`);
        break;
      default:
        acc[item] = params[item];
    }

    return acc;
  }, {});

/**
 * Gets the amount of a given transaction
 *
 * @param {Object} transaction The transaction object
 * @returns {String} Amount in Beddows/Satoshi
 */
const getTxAmount = ({ moduleAssetId, asset }) => {
  if (moduleAssetId === transfer) {
    return asset.amount;
  }

  if (moduleAssetId === unlockToken) {
    return asset.unlockingObjects?.reduce((sum, unlockingObject) =>
      sum + Number(unlockingObject.amount), 0);
  }
  if (moduleAssetId === voteDelegate) {
    return asset.votes.reduce((sum, vote) =>
      sum + Number(vote.amount), 0);
  }

  return undefined;
};

export {
  getTxAmount,
  transformTransaction,
  containsTransactionType,
  createTransactionObject,
  normalizeTransactionParams,
};
