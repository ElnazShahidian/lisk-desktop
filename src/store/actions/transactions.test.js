import { actionTypes } from '@constants';
import * as transactionsApi from '@api/transaction';
import * as awaitToJs from 'await-to-js';
import {
  emptyTransactionsData,
  transactionsRetrieved,
  resetTransactionResult,
  pendingTransactionAdded,
  transactionDoubleSigned,
  transactionBroadcasted,
} from './transactions';

jest.mock('@api/transaction');
jest.mock('@api/delegate');
// jest.mock('await-to-js')
// jest.spyOn(awaitToJs, 'to')
// const toMock = awaitToJs.to
transactionsApi.broadcast = jest.fn(() => {})

describe('actions: transactions', () => {
  const dispatch = jest.fn();
  const getState = () => ({
    network: {
      status: { online: true },
      name: 'Mainnet',
      networks: {
        LSK: {
          serviceUrl: 'http://localhost:4000',
          nethash: '198f2b61a8eb95fbeed58b8216780b68f697f26b849acf00c8c93bb9b24f783d',
        },
      },
    },
    transactions: {
      filters: {},
    },
    settings: {
      token: {
        active: 'LSK',
      },
    },
    blocks: {
      latestBlocks: [{
        timestamp: 123123123,
      }],
    },
  });

  describe('getTransactions', () => {
    const data = {
      address: '15626650747375562521L',
      limit: 20,
      offset: 0,
      filters: {},
    };

    it('should create an action function', () => {
      expect(typeof transactionsRetrieved(data)).toBe('function');
    });

    it('should dispatch getTransactionsSuccess action if resolved', async () => {
      transactionsApi.getTransactions.mockResolvedValue({ data: [], meta: { total: 0 } });
      const expectedAction = {
        count: 0,
        confirmed: [],
        address: data.address,
        filters: data.filters,
        offset: 0,
      };

      await transactionsRetrieved(data)(dispatch, getState);
      expect(dispatch).toHaveBeenLastCalledWith({
        type: actionTypes.transactionsRetrieved,
        data: expectedAction,
      });
    });
  });

  describe('emptyTransaction', () => {
    it('should create an action to empty transactions', () => {
      const expectedAction = { type: actionTypes.emptyTransactionsData };
      expect(emptyTransactionsData()).toEqual(expectedAction);
    });
  });

  describe('pendingTransactionAdded', () => {
    it('should create an action to add pending transaction', () => {
      const data = {
        moduleAssetId: '1938573839:g45krEIjwK',
        id: '4emF3me9YJSbcIuOp',
        fee: '0.001032519',
        nonce: 'wijFKdld1iRd039Bws24UR',
        signatures: ['xnVCm30IUhtYidgBX', 'uxsFGiaqS3n4ydB'],
        sender: {
          address: '3040783849904107057L',
          publicKey: 'MIICXAIBAAKBgQCqGKukO1De7zhZj6+H0qtjTkVxwTCpvKe4eCZ0FPqri0cb2JZfXJ/DgYSF6'
        },
        isPending: true,
      }
      const expectedAction = {
        type: actionTypes.pendingTransactionAdded,
        data,
      };
      expect(pendingTransactionAdded(data)).toEqual(expectedAction);
    });
  });

  describe('resetTransactionResult', () => {
    it('should create an action to reset transaction result', () => {
      const expectedAction = { type: actionTypes.resetTransactionResult };
      expect(resetTransactionResult()).toEqual(expectedAction);
    });
  });

  describe('transactionDoubleSigned', () => {
    it('should create an action to reset transaction result', () => {
      const data = {
        id: 'a66c3cb626dbb631e53a5bd617cc8c13f06186c91f42e87008331e6dc4dc3cba',
        signatures: ['ofh49fhfnYHK8499dvvn', '0egeJT7dgdnek4n5k6994k'],
        senderPublicKey: 'MIICXAIBAAKBgQCqGKukO1De7zhZj6+H0qtjTkVxwTCpvKe4eCZ0FPqri0cb2JZfXJ/DgYSF6',
      }
      const expectedAction = {
        type: actionTypes.transactionDoubleSigned,
        data
      };
      expect(transactionDoubleSigned(data)).toEqual(expectedAction);
    });
  });

  describe('transactionBroadcasted', () => {
    it('should create an action to broadcast transaction', async () => {
      const data = {
        recipientAddress: 'lsktbdkqr8axg3rovndj997dveyrfctrpo95e9byc',
        amount: 149999992,
        fee: 0.001483953,
        dynamicFeePerByte: 0.000746375,
        reference: '',
        // id: 'a66c3cb626dbb631e53a5bd617cc8c13f06186c91f42e87008331e6dc4dc3cba',
        // signatures: ['ofh49fhfnYHK8499dvvn', '0egeJT7dgdnek4n5k6994k'],
        // senderPublicKey: 'MIICXAIBAAKBgQCqGKukO1De7zhZj6+H0qtjTkVxwTCpvKe4eCZ0FPqri0cb2JZfXJ/DgYSF6',
      }
      // toMock.mockResolvedValue(data)
      // awaitToJs.to = jest.fn.mockImplementation(() => Promise.resolve(data))
      // jest.spyOn(awaitToJs, 'to').mockImplementation(async () => data)
      jest.spyOn(awaitToJs, 'to').mockImplementation(async (new Promise(resolve => resolve({}))) => Promise.resolve(data))
      const expectedAction = {
        type: actionTypes.transactionBroadcasted,
        data,
      };
      // transactionsApi.broadcast.mockReturnValue({})
      // console.log(transactionBroadcasted(data))
      await transactionBroadcasted(data)(dispatch, getState)
      // expect(toMock).toHaveBeenCalledTimes(1)
      expect(dispatch).toHaveBeenCalledWith(expectedAction);
    });
  });
});
