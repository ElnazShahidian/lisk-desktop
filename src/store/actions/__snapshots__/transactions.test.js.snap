// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions: transactions transactionBroadcasted should dispatch pendingTransactionAdded action if transformedTransaction sender address is the same as LSK address 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": Object {
          "asset": Object {
            "votes": Array [
              Object {
                "amount": 2000000000n,
                "delegateAddress": Object {
                  "data": Array [
                    185,
                    194,
                    40,
                    186,
                    229,
                    249,
                    166,
                    248,
                    160,
                    189,
                    119,
                    135,
                    180,
                    193,
                    35,
                    191,
                    29,
                    228,
                    190,
                    221,
                  ],
                  "type": "Buffer",
                },
              },
            ],
          },
          "assetID": 1,
          "fee": 142000n,
          "id": "ad0e0acbe8a3ece3087c8362149ca39c470e565d268df32e57de5d3fe2e1ea5c",
          "moduleID": 5,
          "nonce": 2n,
          "senderPublicKey": Object {
            "data": Array [
              32,
              86,
              136,
              73,
              43,
              197,
              45,
              218,
              191,
              220,
              16,
              250,
              119,
              40,
              184,
              188,
              181,
              148,
              42,
              209,
              124,
              104,
              171,
              92,
              32,
              233,
              97,
              83,
              253,
              26,
              198,
              87,
            ],
            "type": "Buffer",
          },
          "signatures": Array [
            Object {
              "data": Array [
                75,
                252,
                10,
                181,
                225,
                179,
                195,
                251,
                28,
                167,
                54,
                42,
                204,
                145,
                119,
                117,
                206,
                99,
                69,
                242,
                106,
                38,
                29,
                89,
                45,
                131,
                198,
                42,
                209,
                86,
                233,
                2,
                33,
                38,
                154,
                59,
                66,
                53,
                22,
                194,
                184,
                76,
                30,
                189,
                210,
                133,
                211,
                191,
                131,
                190,
                104,
                143,
                90,
                21,
                140,
                29,
                175,
                114,
                69,
                254,
                206,
                160,
                53,
                10,
              ],
              "type": "Buffer",
            },
          ],
        },
        "type": "BROADCAST_TRANSACTION_SUCCESS",
      },
    ],
    Array [
      Object {
        "data": 2021-11-12T14:25:58.742Z,
        "type": "TIMER_RESET",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`actions: transactions transactionCreated should dispatch transactionCreatedSuccess action if there are no errors 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "data": Object {
          "asset": Object {
            "amount": 21000000n,
            "data": "",
            "recipientAddress": Object {
              "data": Array [
                138,
                38,
                224,
                248,
                38,
                31,
                60,
                176,
                151,
                116,
                145,
                187,
                66,
                220,
                1,
                35,
                47,
                96,
                170,
                208,
              ],
              "type": "Buffer",
            },
          },
          "assetID": 0,
          "fee": 141000n,
          "id": Object {
            "data": Array [
              159,
              61,
              39,
              83,
              86,
              37,
              71,
              15,
              32,
              214,
              177,
              72,
              207,
              162,
              189,
              253,
              43,
              57,
              238,
              26,
              100,
              87,
              68,
              175,
              85,
              152,
              200,
              185,
              79,
              160,
              223,
              200,
            ],
            "type": "Buffer",
          },
          "moduleID": 2,
          "nonce": 0n,
          "senderPublicKey": Object {
            "data": Array [
              32,
              86,
              136,
              73,
              43,
              197,
              45,
              218,
              191,
              220,
              16,
              250,
              119,
              40,
              184,
              188,
              181,
              148,
              42,
              209,
              124,
              104,
              171,
              92,
              32,
              233,
              97,
              83,
              253,
              26,
              198,
              87,
            ],
            "type": "Buffer",
          },
          "signatures": Array [
            Object {
              "data": Array [
                68,
                152,
                111,
                35,
                216,
                251,
                44,
                220,
                176,
                50,
                7,
                2,
                64,
                241,
                247,
                169,
                57,
                175,
                77,
                30,
                145,
                128,
                54,
                20,
                13,
                93,
                106,
                4,
                6,
                102,
                182,
                227,
                29,
                2,
                205,
                220,
                153,
                88,
                94,
                100,
                220,
                0,
                235,
                246,
                35,
                57,
                220,
                158,
                108,
                169,
                33,
                187,
                43,
                102,
                31,
                207,
                239,
                187,
                4,
                139,
                174,
                87,
                206,
                5,
              ],
              "type": "Buffer",
            },
          ],
        },
        "type": "TRANSACTION_CREATED_SUCCESS",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
